src/pipeline/comms/channel.rs

```diff
@@ -0,0 +1,146 @@
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! // helper function that sets up 3 helpers hooked up to each other with channels
+//! let (c1, c2, c3, c_run) = raw_ipa::pipeline::util::intra_process_comms();
+//! tokio::spawn(c_run); // this initializes all of the runtime pieces for channels
+//!
+//! #[derive(Clone, serde::Serialize, serde::Deserialize)]
+//! struct ExampleRequest {
+//!     message: String,
+//! }
+//! let req = ExampleRequest {
+//!     message: String::from("hello"),
+//! };
+//! c1.send_to(Target::Next, req.clone()).await?;
+//! let recvd = c2.receive_from::<ExampleRequest>(Target::Prev).await?;
+//! assert_eq!(req.message, recvd.message);
+//! # Ok(())
+//! # }
+//! ```
+
+use crate::pipeline::comms::buffer::{self, as_source};
+use crate::pipeline::comms::{Comms, Target};
+use crate::pipeline::Result;
+use async_trait::async_trait;
+use log::{debug, error, info};
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::sync::mpsc;
+use tokio::time;
+use uuid::Uuid;
+
+pub struct Channel<B: buffer::Buffer> {
+    name: &'static str,
+    next_send_chan: mpsc::Sender<Vec<u8>>,
+    prev_send_chan: mpsc::Sender<Vec<u8>>,
+    buffer: B,
+    shared_id: Uuid,
+}
+impl<B: buffer::Buffer> Channel<B> {
+    #[must_use]
+    pub fn new(
+        name: &'static str,
+        next_send_chan: mpsc::Sender<Vec<u8>>,
+        prev_send_chan: mpsc::Sender<Vec<u8>>,
+        buffer: B,
+        shared_id: Uuid,
+    ) -> Channel<B> {
+        Channel {
+            name,
+            next_send_chan,
+            prev_send_chan,
+            buffer,
+            shared_id,
+        }
+    }
+
+    pub async fn receive_data(&self, mut recv_chan: mpsc::Receiver<Vec<u8>>) {
+        while let Some(data) = recv_chan.recv().await {
+            let chan_mess_res: Result<ChannelMessage> =
+                serde_json::from_slice(data.as_slice()).map_err(Into::into);
+            match chan_mess_res {
+                Err(err) => error!("received unexpected message: {}", err),
+                Ok(chan_mess) => {
+                    let sent = self
+                        .buffer
+                        .write(chan_mess.shared_id, chan_mess.source, chan_mess.data)
+                        .await;
+                    if sent.is_err() {
+                        error!("could not send message to buffer: {}", sent.unwrap_err());
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<B: buffer::Buffer> Drop for Channel<B> {
+    fn drop(&mut self) {
+        info!("{} comms closing", self.name);
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct ChannelMessage {
+    shared_id: Uuid,
+    source: Target,
+    data: Vec<u8>,
+}
+
+#[async_trait]
+impl<B: buffer::Buffer + 'static> Comms for Channel<B> {
+    async fn send_to<S: serde::Serialize + Send>(&self, target: Target, data: S) -> Result<()> {
```

Comment:
why does `S` here need to be `Send` @akoshelev ?

Reply:
I believe it happens because for futures to implement `Send`, everything they capture must be `Send` too. `data` is used inside `send_to`, meaning that future state machine must capture it, therefore it must be `Send`. 

In addition to arguments being `Send` if they are used inside function, everything that gets carried across await points must be `Send` too. Anything that is **created** and destroyed between await points does not need to be `Send`. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=60ce0f9c43820b178c86c650fd69e493)

Reply:
thanks, this makes sense